pkg/kubelet/nodestatus/setters.go

```diff
@@ -100,51 +100,19 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 				}
 				if len(enforcedNodeAddresses) > 0 {
 					for _, nodeAddress := range nodeAddresses {
-						if !nodeIPTypes[nodeAddress.Type] && nodeAddress.Type != v1.NodeHostName {
+						if !nodeIPTypes[nodeAddress.Type] {
 							enforcedNodeAddresses = append(enforcedNodeAddresses, v1.NodeAddress{Type: nodeAddress.Type, Address: nodeAddress.Address})
 						}
 					}
 
-					enforcedNodeAddresses = append(enforcedNodeAddresses, v1.NodeAddress{Type: v1.NodeHostName, Address: hostname})
+					enforcedNodeAddresses = ensureHostname(hostname, hostnameOverridden, enforcedNodeAddresses)
```

Comment:
previously, this code was guaranteed to get an address of type hostname. now, there are scenarios where no hostname address is reported:

* nodeIP="1.1.1.1"
* hostname="2.2.2.2"
* hostnameOverride=false
* cloud provider reports:
  * type=internalIP, address=1.1.1.1
  * type=internalIP, address=2.2.2.2

enforcedNodeAddresses filters the internalIP list to "1.1.1.1", then ensureHostname sees no existing address matching "2.2.2.2" (even though the cloud provider reported one) so it avoids appending "2.2.2.2"

Reply:
I think passing `nodeAddresses` rather than `enforcedNodeAddresses` would fix that, but maybe this whole thing could be reorganized to put the `nodeIP` and non-`nodeIP` cases more in-sync?

Reply:
> I think passing nodeAddresses rather than enforcedNodeAddresses would fix that

that would need to be a second parameter (`cloudProviderReportedAddresses`) distinct from the filtered `enforcedAddresses` list.

Reply:
Fixed: I've refactored to 2 functions `shouldUseCloudProviderReportedHostname` and `ensureHostname` and now `nodeAddresses` (cloud provider reported) is passed in and used for the decision making, not the filtered `enforcedAddresses`